### PR TITLE
Dgs 47 /functionality to handle one page checkout prestashop 16

### PR DESCRIPTION
**Pull request by arturas-bes**
As phone number is not saved on order create, as in opc controller there is not form to submit, built ajax request which triggers on change to handle phone number. Added validation message when phone number contains letters, as it stop json parsing and trows error message.

### CODE REVIEW

**Comment by arturas-bes** [comment]
- Ajax request to save phone number on order opc carrier, number validation added
- Validations for empty phone number and invalid number
- Validations when no pudo selected
- Reload map on address country change
- Bugfix for missing pudo points added
- Fix for parcel update added, update all points

**Review by margud — APPROVED** [review]

File: controllers/front/Ajax.php [hunk 1]
```diff
@@ -107,8 +108,25 @@ public function postProcess()
...
-                $countryCode = Configuration::get(Config::WEB_SERVICE_COUNTRY);
                 $this->updateStreetSelect($countryCode, $city);
+                break;
+            case 'saveSelectedPhoneNumber':
+                /** @var  \Invertus\dpdBaltics\Service\CarrierPhoneService $phoneService */
+                $phoneService = $this->module->getContainer(\Invertus\dpdBaltics\Service\CarrierPhoneService::class);
+
+                if(!$phoneService->saveCarrierPhone($cartId,
+                    Tools::getValue('phone_number'),
+                    Tools::getValue('phone_area')
```
Comment (tomas862) [inline controllers/front/Ajax.php:119]: does the service inisde has phone number validation ? Do you validate it on back-end side, if so , how ? If not, then I suppose you should do that.

File: dpdbaltics.php [hunk 2]
```diff
@@ -159,6 +159,18 @@ public function hookActionFrontControllerSetMedia()
...
+
+        if ($currentController === 'order-opc') {
+            $this->context->controller->addJS($this->getPathUri() . 'views/js/front/order-opc.js');
+            Media::addJsDef([
+               'order_opc_errors' => [
+                   'pickup_point_error' => $this->l('Please select pickup point'),
+                   'invalid_phone_error' => $this->l('Invalid phone number'),
+                   'empty_phone_error' => $this->l('Please fill up phone number'),
+                   'invalid_delivery_time' => $this->l('Please select delivery time')
+               ]
```
Comment (tomas862) [inline dpdbaltics.php]: in mollie project we faced issue when French people translating stuff by adding this symbol `'`. Try to add it, you will see that javascript will get broken . The solution to it is to use native php function addSlashes for texts, wrap each text with it and it should be fine - you can test this

File: views/js/front/order-opc.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            'phone_number': phoneNumber,
+            'phone_area': phoneArea,
+            'action': 'saveSelectedPhoneNumber',
+            'token': static_token
+        },
+        success: function (response) {
+        },
+        error: function (response) {
+
+        }
```
Comment (tomas862) [inline views/js/front/order-opc.js]: what happens if I send invalid phone number ?

Comment (tomas862) [reply views/js/front/order-opc.js]: I suppose you should handle both success and error cases ?

File: controllers/front/Ajax.php (hunk 1 above)
Comment (arturas-bes) [reply controllers/front/Ajax.php:119]: Added validation inside function

File: views/js/front/order-opc.js (hunk 3 above)
Comment (arturas-bes) [reply views/js/front/order-opc.js]: I show message in network now does not reach front as I already did front validation not to execute script if invalid Atm when number is invalid i stops click event and backend throws invalid phone number never saves phone

**Review by tomas862 — APPROVED** [review]

**Pull request merged**